### PR TITLE
Clean up CachedResult Dispose lifecycle

### PR DESCRIPTION
**Pull request by max-hoffman**
Dispose should only be called on iterators whose resources are transient. Resources that outlive a single iterator should be Disposed at the end of the query through the Node (not iter). A CachedResult can be safely Disposed after the reseult is transferred to the parent HashLookup. A HashJoin should not be disposed when its parent is an iterative node that can call RowIter again. But if that is mistaken for some yet unforseen edge case, disposing the CachedResult returns it to a state where it will correctly recompute its contents rather than cycling on a nil'd cache where |cacheDone| is left as true.

### CODE REVIEW

File: sql/plan/cached_results.go [hunk 1]
```diff
@@ -69,6 +93,7 @@ func (n *CachedResults) RowIter(ctx *sql.Context, r sql.Row) (sql.RowIter, error
 }
 
 func (n *CachedResults) Dispose() {
+	n.cacheDone = false
```
Comment (zachmu) [inline sql/plan/cached_results.go:96]: This is kind of non-sensical, smells wrong to me.

Dispose should happen at the end of a node's lifecycle, when all rows have been iterated and there is no chance that additional work is going to be performed on it. The fact that you're doing this early from HashLookup doesn't change this equation.

Shouldn't we be setting n.cacheDone when the child iter return EOF? Is it actually cacheInitialized?

**Review by zachmu — COMMENTED** [review]
Better than before but still rubs me the wrong way, see comment

File: sql/plan/cached_results.go (hunk 1 above)
Comment (max-hoffman) [reply sql/plan/cached_results.go:96]: is this just a naming confusion? to me `cacheInitialized` sounds the cache has started being filled, when i intend it to mean "the cache contains all of the rows of an exhausted child iter". I can make it act like the former and maintain its usefulness. The point of this line is to help make the parent idiot-proof, even if the parent does something not sensible this prevents the infinite spinning bug. the CachedResult does not control whether its parent does a sensible thing.

in aggregate i don't think this complexity is necessary, i think we should delete CachedResult and nodes that want caches should just build them in-line

**Review by zachmu — APPROVED** [review]
LGTM discussed in person

**Pull request merged**